File: youtube-fetch/src/video-id.js
/**
 * 从 YouTube URL 或 11 位 video id 解析出 id。
 * @param {string} input
 * @returns {string | null}
 */
export function parseYouTubeVideoId(input) {
  const raw = String(input ?? "").trim();
  if (!raw) return null;

  if (/^[\w-]{11}$/.test(raw)) return raw;

  let url;
  try {
    url = new URL(raw);
  } catch {
    const m = raw.match(/(?:v=|\/embed\/|\/shorts\/|youtu\.be\/)([\w-]{11})/);
    return m?.[1] ?? null;
  }

  const host = url.hostname.replace(/^www\./, "");
  if (host === "youtu.be") {
    const id = url.pathname.slice(1).split("/")[0];
    return /^[\w-]{11}$/.test(id) ? id : null;
  }
  if (host === "youtube.com" || host === "m.youtube.com" || host === "music.youtube.com") {
    const v = url.searchParams.get("v");
    if (v && /^[\w-]{11}$/.test(v)) return v;
    const parts = url.pathname.split("/").filter(Boolean);
    const shortsIdx = parts.indexOf("shorts");
    if (shortsIdx >= 0 && parts[shortsIdx + 1] && /^[\w-]{11}$/.test(parts[shortsIdx + 1])) {
      return parts[shortsIdx + 1];
    }
    const embedIdx = parts.indexOf("embed");
    if (embedIdx >= 0 && parts[embedIdx + 1] && /^[\w-]{11}$/.test(parts[embedIdx + 1])) {
      return parts[embedIdx + 1];
    }
  }
  return null;
}

File: youtube-fetch-py/youtube_fetch/video_id.py
"""YouTube URL → video id."""

from __future__ import annotations

import re
from urllib.parse import urlparse, parse_qs

_ID_RE = re.compile(r"^[\w-]{11}$")
_EMBED_RE = re.compile(r"(?:v=|/embed/|/shorts/|youtu\.be/)([\w-]{11})")


def parse_youtube_video_id(input_str: str | None) -> str | None:
    raw = str(input_str or "").strip()
    if not raw:
        return None
    if _ID_RE.match(raw):
        return raw

    try:
        url = urlparse(raw if "://" in raw else f"https://{raw}")
    except ValueError:
        m = _EMBED_RE.search(raw)
        return m.group(1) if m else None

    host = (url.hostname or "").removeprefix("www.")
    if host == "youtu.be":
        part = url.path.lstrip("/").split("/")[0]
        return part if _ID_RE.match(part) else None

    if host in ("youtube.com", "m.youtube.com", "music.youtube.com"):
        v = parse_qs(url.query).get("v", [None])[0]
        if v and _ID_RE.match(v):
            return v
        parts = [p for p in url.path.split("/") if p]
        for key in ("shorts", "embed"):
            if key in parts:
                idx = parts.index(key)
                if idx + 1 < len(parts) and _ID_RE.match(parts[idx + 1]):
                    return parts[idx + 1]
    return None
